add a blackhole scenario

marten-seemann opened the pull request:
The description probably needs some more work.

marten-seemann commented on sim/scenarios/blackhole/README.md:
```diff
@@ -0,0 +1,42 @@
+
+This scenario can be used to test various features of QUIC:
+* Short durations (around 3 RTTs) can be used to test PTOs.
+* Slightly longer durations (> 3 RTT + max_ack_delay) can be used to test
```
@janaiyengar Please carefully check these numbers.

janaiyengar replied:
I would remove these numbers from the description, they're not necessary, and can change in the future.

janaiyengar commented on sim/scenarios/blackhole/README.md:
```diff
@@ -0,0 +1,42 @@
+* `--queue`: Queue size of the queue attached to the link. Specified in
+  packets. This is a required parameter. For example `--queue=25`.
+
+* `--on`: Time that traffic is allowed to flow. Specify with units. This is a
```
```suggestion
* `--on`: Time period that traffic is allowed to flow. Specify with units. This is a
```

janaiyengar commented on sim/scenarios/blackhole/README.md:
```diff
@@ -0,0 +1,42 @@
+* `--on`: Time that traffic is allowed to flow. Specify with units. This is a
+  required parameter. For example `--on=10s`.
+
+* `--off`: Time that traffic is blocked. Specify with units. This is a required
```
```suggestion
* `--off`: Time period that traffic is blocked. Specify with units. This is a required
```

janaiyengar commented on sim/scenarios/blackhole/blackhole-error-model.cc:
```diff
@@ -0,0 +1,27 @@
+ 
+BlackholeErrorModel::BlackholeErrorModel() : enabled_(true) { }
+ 
+bool BlackholeErrorModel::DoCorrupt(Ptr<Packet> p) {
```
does this cause the packet to get dropped or corrupted?

marten-seemann replied:
It just gets dropped. You could corrupt the packet by modifying the raw packet contained in `p`.

janaiyengar commented on sim/scenarios/blackhole/blackhole.cc:
```diff
@@ -0,0 +1,77 @@
+
+  int repeat = 1;
+  if(repeat_s.length() > 0) {
+    repeat = std::stoi(repeat_s);
```
I would suggest adding a NS_ABORT_MSG_IF(repeat <= 0, "Invalid valid: repeat value must be greater than zero.").

marten-seemann replied:
Makes sense. Applied.

janaiyengar commented on sim/scenarios/blackhole/blackhole.cc:
```diff
@@ -0,0 +1,77 @@
+  std::cout << Simulator::Now().GetSeconds() << "s: Disabling blackhole" << std::endl;
+  counter++;
+  em->Disable();
+  if(repeat > 0 && counter < repeat) {
```
You can remove this repeat > 0 check here if you take my suggestion below.

janaiyengar commented on sim/scenarios/blackhole/blackhole.cc:
```diff
@@ -0,0 +1,77 @@
+
+  Ptr<BlackholeErrorModel> em = CreateObject<BlackholeErrorModel>();
+  em->Disable();
+  devices.Get(0)->SetAttribute("ReceiveErrorModel", PointerValue(em));
```
Does this apply to traffic in both directions then? It would be nice to have the directions be independent.

marten-seemann replied:
It actually only drops packets in the server -> client direction.
For the other direction, we'd have to use the same error model on `devices.Get(1)` as well I guess. I'll add a command line option for that.

marten-seemann commented:
@janaiyengar The code was only blocking packets in the direction from the server to the client. I modified it such we're blocking both directions by default, and added a command line flag to configure that.
